CRM/Contribute/BAO/Contribution.php

```diff
@@ -233,6 +233,10 @@ public static function add(&$params, $ids = array()) {
     $params['contribution'] = $contribution;
     self::recordFinancialAccounts($params);
 
+    if (!$contributionID) {
+      self::storeInvoiceNumber($contribution->id);
+    }
```

Comment:
set invoice number when new contribution is created.

Reply:
We could prevent an extra api.contribution.create call by generating the invoice number on line 225 and saving that field with the rest. Granted you need an id for this, but you could query the db for the next available id which is pretty reliable if I'm not mistaken.

Reply:
OR what if we call ```CRM_Core_DAO::setFieldValue(__CLASS__, $contributionID, 'invoice_number', $invoiceNumber);``` instead of api.contribution.create since the changes is in BAO ?

Reply:
I agree with Coleman re: looking up max(invoice_number) and setting value before record is stored in order to avoid second write of contribution record. But we need to make sure that we have this in a transaction that will prevent two contributions from getting the same incremented invoice_number